lib/portage/gpg.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2001-2020 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+import subprocess
+import sys
+import threading
+import time
+
+from portage import os
+from portage.exception import GPGException
+from portage.output import colorize
+from portage.util import shlex_split, varexpand, writemsg, writemsg_stdout
+
+class GPG:
+	"""
+	Unlock GPG, must call dircetly from main program for get correct TTY
+	"""
+	def __init__(self, settings):
+		"""
+		Portage settings are needed to run GPG unlock command.
+		"""
+		self.settings = settings
+		self.GPG_unlock_command = self.settings.get(
+			"BINPKG_GPG_UNLOCK_COMMAND", None)
+		if "gpg-keepalive" in self.settings.features:
+			self.keepalive = True
+		else:
+			self.keepalive = False
+
+	def unlock(self):
+		"""
+		Set GPG_TTY and run GPG unlock command.
+		If gpg-keepalive is set, start keepalive thread.
+		"""
+		if (self.GPG_unlock_command
+			and (self.settings.get("BINPKG_FORMAT", "xpak") == "gpkg")):
+			try:
+				os.environ["GPG_TTY"] = os.ttyname(sys.stdout.fileno())
```

Comment:
Same as tar_stream_reader comment about putting GPG_TTY in the Popen env.

Reply:
This one is special since it call os.system() and give the tty to GPG for ask password. But os.system() does not have something likes Popen env. So I have to add it to current environment.

Reply:
We should be able to use Popen here instead of os.system here, like you do in the gpg_keepalive method. You just need to do the shlex_split earlier, and wait for the process like this:
```
return_code = subprocess.Popen(self.GPG_unlock_command).wait()
```

Reply:
I'd like to, but how can I pass the stdin to gpg? I tried to pass the current stdin as a parameter, but it didn't work at all.

Reply:
If stdin is unspecified or None, then the Popen process inherits stdin automatically.

Reply:
But it not works in GPG, pinentry read in a special way. 
When using pinentry-curses, it has no output.
--pinentry-mode loopback need wait like 30s to get the prompt.

Reply:
I've tested locally, and it seems to work for me with either pinentry-curses or loopback mode as long as I have `Defaults use_pty` in /etc/sudoers. I think for this kind of usage it's probably best to enable use_pty in sudoers and not export GPG_TTY (GPG_TTY is probably never a good idea, see https://dev.gnupg.org/T3412).

Reply:
And the user has to be the root, which probably fine for portage. I am doing to test that.